File: leetcode/189RotateArray.java
class Solution {
    public void rotate(int[] nums, int k) {
        int[] result = new int[nums.length];
        
        for(int i=0;i<nums.length;i++){
            int newIndex = (i+k)%nums.length;
            
            result[newIndex] = nums[i];
        }
        
        for(int i=0;i<nums.length;i++){
            nums[i] = result[i];
        }
    }
}

// O(n) space

File: leetcode/189RotateArray.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k = k % n
        def reverse(start, end):
            while start < end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1
        reverse(0, n-k-1)
        reverse(n-k, n-1)
        reverse(0, n-1)
